sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java

```diff
@@ -572,15 +599,165 @@ void updateDAI_should_not_update_DAI_values_when_not_updatable() {
+        assertThat(dai.getVal().get(0).getValue()).isEqualTo(NEW_VALUE);
+    }
+
+    @Test
+    void updateDAI_with_SDO_and_BDA_when_only_DOI_is_instantiated_should_create_DOI_SDI_DAI() {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        AbstractLNAdapter<?> lnAdapter = findLn0(new SclRootAdapter(scd), "IED_NAME", "LD_INS1");
+        lnAdapter.getDOIAdapterByName(DO_TYPE_NAME.getName()).getCurrentElem().getSDIOrDAI().clear();
+
+        System.out.println("BEFORE => " + MarshallerWrapper.marshall(scd));
+
+        // When
+        lnAdapter.updateDAI(RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3);
+
+        // Then
+        System.out.println("AFTER => " + MarshallerWrapper.marshall(scd));
+        TDAI dai = findDai(lnAdapter, RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3.getDataAttributes()).getCurrentElem();
+        assertThat(dai.isValImport()).isTrue();
+        assertThat(dai.getVal().get(0).getValue()).isEqualTo(NEW_VALUE);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"sdo1", "d", "antRef", "bda1", "bda2"})
+    void updateDAI_with_SDO_and_BDA_when_some_SDI_instantiated_should_create_missing_SDI_DAI(String lastSdiToKeep) {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        AbstractLNAdapter<?> lnAdapter = findLn0(new SclRootAdapter(scd), "IED_NAME", "LD_INS1");
+        DOIAdapter doi = lnAdapter.getDOIAdapterByName(DO_TYPE_NAME.getName());
+        Iterator<String> sdiNames = SDI_NAMES.iterator();
+        IDataParentAdapter sdiToKeep = doi.getStructuredDataAdapterByName(sdiNames.next());
+        while (!lastSdiToKeep.equals(sdiToKeep.getName())) {
+            sdiToKeep = sdiToKeep.getStructuredDataAdapterByName(sdiNames.next());
+        }
+        assertThat(sdiToKeep.getName()).isEqualTo(lastSdiToKeep);
+        sdiToKeep.getSDIOrDAI().clear();
+
+        System.out.println("BEFORE => " + MarshallerWrapper.marshall(scd));
+
+        // When
+        lnAdapter.updateDAI(RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3);
+
+        // Then
+        System.out.println("AFTER => " + MarshallerWrapper.marshall(scd));
+        TDAI dai = findDai(lnAdapter, RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3.getDataAttributes()).getCurrentElem();
+        assertThat(dai.isValImport()).isTrue();
+        assertThat(dai.getVal().get(0).getValue()).isEqualTo(NEW_VALUE);
+    }
+
+    @Test
+    void updateDAI_with_SDO_and_BDA_when_only_DAI_is_missing_should_create_DAI() {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        AbstractLNAdapter<?> lnAdapter = findLn0(new SclRootAdapter(scd), "IED_NAME", "LD_INS1");
+        AbstractDAIAdapter<?> daiToRemove = findDai(lnAdapter, DATA_TYPE_REF);
+        ((IDataParentAdapter) daiToRemove.getParentAdapter()).getSDIOrDAI().clear();
+
+        System.out.println("BEFORE => " + MarshallerWrapper.marshall(scd));
```

Comment:
Before and after prints are useless.
You can use something similar to this instead. 
```java
// Given
...
var before = lnAdapter.findDoiAdapterByName(DO_TYPE_NAME.getName()).get()
        .getStructuredDataAdapterByName("sdo1")
        .getStructuredDataAdapterByName("d")
        .getStructuredDataAdapterByName("antRef")
        .getStructuredDataAdapterByName("bda1")
        .getStructuredDataAdapterByName("bda2");
assertThat(before.getSDIOrDAI().size()).isEqualTo(0);
// When
lnAdapter.updateDAI(RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3);
// Then
assertThat(before.getSDIOrDAI().size()).isEqualTo(1);
var expected = before.getDataAdapterByName("bda3");
assertThat(expected.getCurrentElem().getVal().get(0).getValue()).isEqualTo(NEW_VALUE);
...
```

Reply:
OK. I removed the print BEFORE and AFTER and added an assertion at the end of the  `//Given` part